crates/relayer/src/util/profiling.rs

```diff
@@ -0,0 +1,137 @@
+use core::sync::atomic::AtomicUsize;
+use core::sync::atomic::Ordering::Relaxed;
+use once_cell::sync::OnceCell;
+use serde_derive::Serialize;
+use serde_json::Value;
+use std::fs::File;
+use std::fs::OpenOptions;
+use std::path::Path;
+
+std::thread_local! {
+    pub static DEPTH: AtomicUsize = AtomicUsize::new(0);
+}
+
+static FILE: OnceCell<File> = OnceCell::new();
+static ENABLED: OnceCell<bool> = OnceCell::new();
+
+pub fn enable() {
+    ENABLED.set(true).expect("profiling already enabled");
+}
+
+pub fn enabled() -> bool {
+    ENABLED.get().copied().unwrap_or(false)
+}
+
+/// Measure the time between when this value is allocated
+/// and when it is dropped.
+pub struct Timer {
+    name: &'static str,
+    info: Value,
+    start: std::time::Instant,
+}
+
+#[derive(Debug, Serialize)]
+struct TimerInfo<'a> {
+    name: &'a str,
+    #[serde(flatten)]
+    info: &'a Value,
+    elapsed: u128,
+}
+
+impl Timer {
+    pub fn new(name: &'static str, info: Value) -> Self {
+        let depth = DEPTH.with(|d| d.fetch_add(1, Relaxed));
+        let pad = "   ".repeat(depth);
+
+        if enabled() {
+            tracing::info!("{}⏳ {} - start", pad, name);
+        }
+
+        Self {
+            name,
+            info,
+            start: std::time::Instant::now(),
+        }
+    }
+}
+
+impl Drop for Timer {
+    fn drop(&mut self) {
+        let enabled = enabled();
```

Comment:
```suggestion
        if !enabled() {
            return;
        }
        let enabled = enabled();
```
What about skipping everything if profiling isn't enabled, do you see any downside to this ?

Reply:
https://github.com/informalsystems/hermes/pull/3352/commits/bc105a818c9a38b9b8c6bdd0f010c85fd1a6b5e2